### ISSUE TITLE
SPFx AADHttpClient post stalled for minutes by the browser

### ISSUE THREAD

**Issue by spplante** [SharePoint/sp-dev-docs#3003]
#### Category
- [ ] Question
- [ ] Typo
- [x] Bug
- [ ] Additional article idea

#### Expected or Desired Behavior

When using **AadHttpClient.post()**, the response should come back right after the *Azure* endpoint is done working.

#### Observed Behavior

I built an SPFx extension which uses the **AadHttpClient** to execute a POST request to an Azure Function from our tenant. 

The extension authenticates to the Azure Function just fine, the *Azure Function* takes about 2-5 seconds to execute, but the JavaScript code hangs for sometimes up to 1-2 minutes before entering the **.then()** of the initial post()...

Sometimes everything goes though normally but sometimes, in a random fashion, the client side JavaScript gets stalled during 1-2 minutes...I did verify the execution time on the Azure side, always 2-5 seconds. I get the problem both in IE and Chrome.

Looking at the **Network** tab, I get something really strange, you can see the initial call to my Azure Function **ApplyFolderTemplate**, which gets stalled during **1.1 minutes** because of all the other calls being executed by **spopushchannel-mini-1029a99e.js** to the following URL :

**https://1-canadaeast1-pushp.svc.ms/socket.io/?token=xxxxxxxxxx&EIO=3&transport=polling&t=MT1v_DS&sid=xxxxxxxxx**

![Network Tab](https://image.ibb.co/hvxQFV/Issue.png)

When inspecting the original post which took 1.1 minute, Chrome shows the following information :

![Network Tab Inspection](https://image.ibb.co/kno7vV/Issue2.png)

Each of these lines are being added one at the time, taking multiple seconds to resolve, to finally resolve with a 200 status and start a new line exactly like the previous one, all this during 1-2 minutes. It looks like something is wrong in the authentication mechanism but at this point I am not sure what.

I tried to take a look at the code but I am not doing much, all I do is a simply AadHttpClient.post() with no particular headers or anything...

#### Steps to Reproduce

- Create an SPFx solution which executes a post call to an Azure Function using the AadHttpClient
- Inspect the network tab to see the client side processing time vs server side processing time

**Comment by patmill** [comment]
Is this running on a list page?  If so, does the same thing happen on a webpart page?

**Comment by spplante** [comment]
Yes it is a list extension so right now I can't confirm if the same problem occurs on a WebPart page.

**Comment by patmill** [comment]
Looking a bit more at your picture, the stall and the socketio calls don't seem to line up correctly.  Additionally, the socketio calls are sequential, so I wouldn't expect them to do things like consume 6 ports, etc.

Can you post a longer span of network traffic - particularly what is happening _before_ the apply template call?  Also, I assume that the apply_template call is your own Azure function.  Does it happen with other graph calls (like getting the current users manager, etc.)?

**Comment by spplante** [comment]
I can get back to you with a longer network screenshot, as for the "before", the ApplyTemplate is called on a button click so there isn't really anything relevant in the "before", everything you see the screenshot starts from the moment the button is clicked (and the POST executed). I could try it out with Graph just to be sure it has nothing to do with my Function that would be a good hint.

Not sure what could be wrong with my Function though, calling the function through C# doesn't do this, only when using the AadHttpClient.post

**Comment by spplante** [comment]
So I did it twice again and the result is always the same :

![First Time](https://i.ibb.co/QXQmhwD/Screenshot1.png)

![SecondTime](https://i.ibb.co/GvfFXnG/Screenshot2.png)

Both times, the 25 seconds delay apears exactly 3 times and the main call gets stalled exactly 1.1 min.

After I get the response the network tab stops moving.

Although, If I wait long enough, I continue getting the 25 seconds calls as we can see here :

![ThirdTime](https://i.ibb.co/Y7Kqn6R/Screenshot3.png)

The fact that it keeps going even though the response came back makes me think that these 25 seconds calls might not be related at all to the main call being stalled during 1.1 min?

**Comment by patmill** [comment]
Right - the 25 second socketio request is just the library app polling for changes to the backing list being displayed, and the server responds after 25 seconds of no changes.   Looking at your images, is there a reason why the applyTemplate with an OPTIONS method stalls, while a call with a POST wouldn't?

**Comment by spplante** [comment]
Oh ok makes sens.

As for the OPTIONS call, correct me if I am wrong, but even though I am executing a POST in the code, the browser executes an OPTIONS because of the preflight check which has something to do with CORS (which is enabled on my function). I told myself I could try doing a "fetch" instead of a "post" and manually specify the "post" method, but looking at the source code of PNP this is exactly what the post() does, so I guess I don't have any control on this, or do I?

**Comment by lahuey** [comment]
@spplante you are correct. The browser is responsible for making the OPTIONS call. You don't have any control over it.

Please reopen this issue if there are any additional questions. 
Thanks!

**Comment by spplante** [comment]
Not sure why this is getting closed as this doesn’t explain the issue with the browser stalling a simple POST when executed with the AADHttpClient...

**Comment by lahuey** [comment]
@spplante could you email me at lahuey@microsoft.com with your tenant's domain and the sprequestguid header from a stalled request? 

Thanks!

**Comment by spplante** [comment]
Will do tomorrow morning at the office! Thanks!

**Comment by spplante** [comment]
I am closing this issue, after more investigation on our side it seems like my client has a much larger network problem. 

Because of some SSL Inspections, some URLs are randomly not answering and this is for anything O365 related. We discovered lately that laptops using the wifi network are NOT having the problem, while laptops connected to the network by wire do have some problems because of SSL Inspections issue and fail to open some URLs (which is something Microsoft recommends to turn off with O365).

After discovering this I tested my endpoint using the WIFI and I couldn't reproduce the issue, while I could reproduce it when I was wired, so it is most likely part of our network problem.

Sorry about that, didn't expect this to be such a deep problem 👍
